[REG-2184] Initial docs for Bot Actions and End Criteria (#102)
Regression-Games/RegressionDocs: docs/core-concepts/bot-sequences/actions.mdx

---
sidebar_label: 'Bot Actions'
---

# Bot Actions

Bot Actions are the core pieces of logic that drive behavior within a segment of your test sequence. Actions
can range from simple operations like clicking or keypresses, to more complex operations like clicking on
an object in the scene using computer vision.

## Structure of a Bot Action

Bot Actions are defined in Bot Segments within the `botAction` field. A segment may only contain one bot action.

```json
// Segment structure
{
    "name": "Click the play button",
    "description": "Clicks the play button in the main menu",

>     "keyFrameCriteria": [...],

Reviewer: should say 'endCriteria'

    "botAction": {
        "type": ..., // The type of action to be taken, as a string
        "data": {
            ... // The data for the action, as a JSON object
        }
    }
}
```
A bot action has a `type` and a `data` field. The `type` indicates the type of action to be taken, and the `data` field contains
the parameters for that action. For example, a bot action to click on a coordinate may look like this:

```json
{
    "name": "Click the play button",
    "description": "Clicks the play button in the main menu",

>     "keyFrameCriteria": [...],

Reviewer: should say 'endCriteria'

* `MonkeyBot`: This action runs a chaos monkey bot for a given amount of time.
* `Mouse_CVImage`: This action clicks on a region of the screen based on a template image found using computer vision.
* `Mouse_CVText`: This action clicks on a region of the screen based on text found using computer vision.
* `Mouse_ObjectDetection`: This action clicks on a region of the screen based on an object found using computer vision.
* `RestartGame`: This action restarts the game.
* `QuitGame`: This action quits the game.

### InputPlayback

The `InputPlayback` action type plays back a sequence of mouse and keyboard inputs. This can be useful for
replaying a sequence of static actions that were previously recorded as part of a play through.

**Data Format**

`type` is set to `InputPlayback`.

```json
// The `botAction.data` format
{
    "startTime": 0.0, // The time offset at which the action should be performed from the start of the segment
    "inputData": {
        "mouse": [ // Any number of mouse inputs can be provided
            {
                "startTime": ..., // A number indicating the time offset at which the action should be performed from the start of the segment (in seconds)

>                 "screenSize": {"x": ..., "y": ...}, // A width and height indicating the size of the screen, which is used to scale the position of the mouse when the screen size is different from the screen size at the time the inputs were recorded

Reviewer: should say 'linearly' scale so it is clear what kind of scaling we do

                "position": {"x": ..., "y": ...}, // A coordinate indicating the position to click, where X and Y are integers
                "worldPosition": null, // An optional coordinate indicating the position to click in world space, which is often null.
                "leftButton": ..., // A boolean indicating whether the left mouse button should be pressed
                "middleButton": ..., // A boolean indicating whether the middle mouse button should be pressed
                "rightButton": ..., // A boolean indicating whether the right mouse button should be pressed
                "forwardButton": ..., // A boolean indicating whether the forward mouse button should be pressed
                "backButton": ..., // A boolean indicating whether the back mouse button should be pressed
                "scroll": {"x": ..., "y": ...}, // An optional scroll amount, where X and Y are integers

>                 "clickedObjectNormalizedPaths": [...], // An optional list of normalized paths to the objects that were clicked

Reviewer: We should add to the comment that these are used to adjust click coordinates when there are variances in resolution scaling or object position from run to run

    "endCriteria": [...],
    "botAction": {
        "type": "InputPlayback",
        "data": {
            "startTime":0.0,
            "inputData": {
                "keyboard": [

                    {"startTime":0.0,"binding":"<Keyboard>/leftShift"},
                    {"startTime":0.0,"endTime":0.0,"binding":"<Keyboard>/h"},
                    {"endTime":0.0,"binding":"<Keyboard>/leftShift"},

                    {"startTime":0.0,"endTime":0.0,"binding":"<Keyboard>/e"},
                    {"startTime":0.0,"endTime":0.0,"binding":"<Keyboard>/l"},
                    {"startTime":0.0,"endTime":0.0,"binding":"<Keyboard>/l"},
                    {"startTime":0.0,"endTime":0.0,"binding":"<Keyboard>/o"}

                ],
                "mouse":[]
            }
        }
    }
}
```

**Example: Moving and clicking the mouse**

```json
{
    "name": "Move and click the mouse",
    "description": "Moves the mouse to a position and clicks it",
    "endCriteria": [...],
    "botAction": {
        "type": "InputPlayback",
        "data": {
            "startTime": 0.0,
            "inputData": {
                "keyboard": [],
                "mouse": [

>                     {"startTime":0.0,"screenSize":{"x":1920,"y":1080},"position":{"x":942,"y":500},"worldPosition":null,"leftButton":false,"middleButton":false,"rightButton":false,"forwardButton":false,"backButton":false,"scroll":{"x":0,"y":0},"clickedObjectNormalizedPaths":["UI Canvas/Start Button"]},

Reviewer: it is good that this is here, we should add a comment stating that it is best practice to have a mouse position input before the click input .  Without this, Unity's input handling can cause some un-expected behaviour with the click especially when the prior mouse position was over a different input component in the UI.

>                     {"startTime":0.1,"screenSize":{"x":1920,"y":1080},"position":{"x":942,"y":500},"worldPosition":null,"leftButton":true,"middleButton":false,"rightButton":false,"forwardButton":false,"backButton":false,"scroll":{"x":0,"y":0},"clickedObjectNormalizedPaths":["UI Canvas/Start Button"]}

Reviewer: we should include a third entry here with the leftButton:false to show the click release

                ]
            }
        }
    }
}
```

### Behaviour

The `Behaviour` action type runs a MonoBehaviour. This can be useful for running agent code that relies on
particular data and information from the code of your game, such as tapping into an existing navigation system.

**Data Format**

`type` is set to `Behaviour`.

```json
// The `botAction.data` format
{

>     "behaviourFullName": "...", // The full class path of the MonoBehaviour to run

Reviewer: fully qualified

    "maxRuntimeSeconds": ... // The maximum amount of time to run the behaviour for in seconds before it is stopped
}
```

The Behaviour itself needs to be defined in the codebase of your game where it can be found during runtime. It should be
a class that inherits from `MonoBehaviour`. In order to terminate the behaviour, you can use `Destroy(this)` to destroy the
behaviour instance.

**Example: Running a pathfinding behaviour for at max 30 seconds**

> 

Reviewer: We need to add a blurb about IRGBotSegmentActionBehaviour to make Behaviours pausible.

We also need to add a short blurb explaining that the Update loop of your behaviour will run as normal until you destroy it or the timer limit is reached at which point the SDK will destroy it

                {
                    if (!agent.hasPath || agent.velocity.sqrMagnitude == 0f)
                    {
                        Destroy(this);
                    }
                }
            }
        }
    }
}
```

```json
{
    "name": "Run pathfinding behaviour",
    "description": "Runs a pathfinding behaviour to navigate to a position",
    "endCriteria": [...],
    "botAction": {
        "type": "Behaviour",
        "data": {
            "behaviourFullName": "RG.Example.NavigateToPosition",
            "maxRuntimeSeconds": 30.0
        }
    }
}
```

### MonkeyBot

The `MonkeyBot` action type runs a chaos monkey bot for a given amount of time. This can be useful for testing the robustness of your game.
This Monkey Bot will randomly select actions to perform from a list of possible actions, and will perform them at a given interval. These
action are found via a code analysis of the game, which must be run before you use this segment type.

**Setup**

In order to setup the monkey bot, you need to run a code analysis of the game. This can be done from the **Regression Games > Configure Bot Actions** menu.

> Once you click that menu option, a panel will appear. Click the "Analyze Actions" button, which will perform an analysis of your game and

Reviewer: We should add a sentence about them filtering out actions.

Author: Added a note about this. The experience here is actually pretty bad.... since Monkey Bots are configured in the botAction JSON now, these settings in the panel are completely ignored (and this is a low hanging fruit I wanted to address soon).

find all the possible actions that can be taken. Once this is done, you can use the monkey bot in your bot segments.

**Data Format**

`type` is set to `MonkeyBot`.

```json
// The `botAction.data` format
{
    "actionInterval": ... // The interval between actions in seconds,
    "actionSettings": {
        "DisabledActionPaths": [
            ... // An optional list of strings of the format "Game Object Path/Method Path" to disable from the monkey bot
        ]
    }
}
```

**Example: Running a monkey bot to click on random objects while ignoring the quit actions**

```json
{
    "name": "Run monkey bot",
    "description": "Runs a monkey bot to click on random objects while ignoring the quit actions",
    "endCriteria": [...],
    "botAction": {
        "type": "MonkeyBot",
        "data": {
            "actionInterval": 0.5,
            "actionSettings": {
                "DisabledActionPaths": [

>                     "Unity UI/Button/Unity.Multiplayer.Samples.BossRoom.Visual.UISettingsCanvas.OnClickQuitButton",

Reviewer: We aren't giving them any context for how to get these strings, we need to add / explain that

Author: Now covered in the note added above, and also added a comment to the JSON.

    - A resource:// path to a READABLE Texture2D in one of your project's Resources folders.
    - A resource:// path to a .bytes TextAsset in one of your project's Resources folders that is a JPG or PNG file saved with a .bytes file extension.  This can be used if you do not want Unity to import your image as a Texture.
- `withinRect` - An optional (can be null/undefined) field to limit the search area to a specific pixel region of the current frame.  The SDK will linearly tranform the supplied `rect` to fit the current resolution using the `screenSize` as the initial reference resolution.
    - `screenSize` - The reference resolution in pixels which defines the screen space that `rect` is defined within.
    - `rect` - The position (x=0, y=0 is bottom left) and size (width, height) of the rectangle that must contain the supplied image data.  The values are defined in pixels.
- `actions` - The list of mouse actions to take at the center point of the returned `rect`.

**Notes**

The `CVImage` type is still in an experimental phase and may provide inconsistent or unexpected results in many situations.

> - Multiple matches of the specified image within a frame may provide incorrect or inconsistent result bounds.

Reviewer: We need to add a blurb about how images are compared in grey scale

Author: Added

- False positives or incorrect result bounds may occur if pixel regions similar to the specified image exist within a frame.

**How to get the data for the `imageData` field as a base64 encoded string**

1. Use a screenshot tool to capture the target area of your game and save this as a JPG file (PNG/BMP/etc are NOT supported at this time).
2. Run one of the following commands to get the base64 encoded string of the image using the scripts in [this directory](https://github.com/Regression-Games/RGUnityBots/tree/main/user-tools).

```shell
CURRENT_PATH> python encode_jpg_base64.py sample_images/tree.jpg
/9j/4AAQ...<rest of the base64 encoded image data>...VLj3ieS/9k=
```
or
```shell
CURRENT_PATH> ./encode_jpg_base64.sh sample_images/tree.jpg
/9j/4AAQ...<rest of the base64 encoded image data>...VLj3ieS/9k=
```

**Example: Using base64 byte[]**
```json
{
    "name": "CV Image Action: Click Menu Change Profile Button",
    "description": "Moves the mouse over the change profile button, then clicks and releases on the button. Criteria waits for the action to complete.",
    "endCriteria": [

>         {"type":"ActionComplete", "transient":false, "data":{}}

Reviewer: This is the first time `endCriteria` has been filled in.  Either exclude this everywhere (Eesh) or fill this in everywhere and provide a quick explainer at the start.

Author: Good catch, forgot to remove during the copy paste. Fixed!

            ]
        }
    }
}
```

### Mouse_CVText

This action type clicks on a region of the screen based on where the specified text is found. This is useful for
clicking on certain buttons or fields that may have dynamic positions or are changing often during development.

**Data Format**

`type` is set to `Mouse_CVText`.

- `text` - The text to find.  Note that this algorithm will treat this as a set of words to find, NOT a sentence.  Thus if you have 'Create New Profile' on a single line on one button, but have 'Create' 'New' 'Profile' each on a separate line in different screen area, the algorithm has to choose which one to select.  The current method is to select the smallest bounding area containing all required words that is also `withinRect` (if specified).
- `textMatchingRule` - One of `Matches` or `Contains`
    - `Matches` - Each word in the provided `text` must be found as a whole word in the results.  This is a very exact matching rule and can sometimes lead to inconsistent results in frames with poor in game lighting on the text or low text contrast.

>     - `Contains` - Each word in the provided `text` must be found as a part of a word in the results.  This is a looser matching rule and if often used instead of `Matches` for more stable results.  For example "Time" would match to "Timer", "Time", "Times", "Timed", "Time", etc in the frame.  This may not seem ideal, but in most game situations, the contrast or text layout isn't always clearly identifiable and extra or incorrect letters may consistently be found for the text you are looking for in a specific frame of the game.  When you encounter those situations, you can adjust your game to give better text clarity/contrast to your users and/or you can utilize `Contains` instead of `Matches`.

Reviewer: "Time" is in the list 2x

Author: Fixed

    "description":"Moves the mouse over the tree, then clicks and releases on the tree. Criteria waits for the action to complete.",
    "endCriteria": [
         {"type":"ActionComplete", "transient":false, "data":{}}
    ],
    "botAction":{    
        "type":"Mouse_ObjectDetection",
        "data": {
            "textQuery": "Tree",
            "threshold": 0.4,
            "actions": [
                {"leftButton":false, "duration":0.1},  // Move the mouse to the center of the tree
                {"leftButton":true, "duration":0.1},   // Click the left mouse button
                {"leftButton":false, "duration":0.1}   // Release the left mouse button
            ]
        }
    }
}
```

### RestartGame

This action type restarts the game. This should only be used as the 'last' segment in your sequence or segment list, and is useful for getting your game
ready for another test.

There is an interface you can implement to override the default behaviour of restarting the game. If this interface is not implemented when 
the `RestartGame` Bot Action is used in a build of your game **outside the UnityEditor**, then the default action will be to call 

> `SceneManager.LoadScene(sceneBuildIndex: 0, mode: LoadSceneMode.Single)`.  **In the UnityEditor**, the action will **always** be to stop and re-enter play mode;

Reviewer: We also destroy all 'dontdestroyonload' game objects before loading scene 0

Author: Added

                        "position": {"x": 907, "y": 796}, // This is the location to click
                        "worldPosition": null,
                        "leftButton": true, // This indicates that the left mouse button should be pressed
                        "middleButton": false,
                        "rightButton": false,
                        "forwardButton": false,
                        "backButton": false,
                        "scroll": {"x": 0, "y": 0},
                        "clickedObjectNormalizedPaths": []
                    }
                ]
            }
        }
    }
}
```

:::info
There are additional fields that can be set on a bot action and many of these `inputData` fields, such as an apiVersion. You can see the full
specification in the [source code](https://github.com/Regression-Games/RGUnityBots/blob/cd68ed0f4b10567eb652477717338d92b4f4571e/src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotAction.cs).
:::

## Action Types

There are many different types of actions that can be used in a bot segment, each useful in their own way.
The list of supported types are as follows:

> * `InputPlayback`: This action plays back a sequence of mouse and keyboard inputs.

Reviewer 2: ```suggestion
* `InputPlayback`: This action plays back a set of mouse and keyboard inputs.
```

Author: Updated, changed to "series" though to at least give the sense that ordering does matter (in the case where the start times are the same)

                "scroll": {"x": ..., "y": ...}, // An optional scroll amount, where X and Y are integers
                "clickedObjectNormalizedPaths": [...], // An optional list of normalized paths to the objects that were clicked. This is used to adjust the click coordinates when there are variances in resolution scaling or object position between runs.
            }
        ],
        "keyboard": [ // Any number of keyboard inputs can be provided
            {
                "startTime": ..., // A number indicating the time offset at which the action should be performed from the start of the segment (in seconds)
                "endTime": ..., // A number indicating the time offset at which the action should be stopped from the start of the segment (in seconds)
                "binding": ..., // A sting representing the binding to be pressed
            }
        ]
    }
}
```

**Notes**

- You can use the `startTime` and `endTime` fields to control the timing of actions. For instance, you can hold shift and press a key to result in a capital letter.
- If two actions have the same `startTime`, the action that appears earlier in the list will be performed first.

> - You can omit an `endTime` of a keyboard action to have the action continue until it is disabled using using an action that only sets the `endTime`.

Reviewer 3: ```suggestion
- You can omit an `endTime` of a keyboard action to have that action continue until it is disabled by a future action which only sets the `endTime`.
```

Honestly still confusing to read, even after the suggestion. I don't want to spend too much time making it read well though unless we think it would be highly useful.

            isPaused = true;
        }

        public void UnPauseAction() {
            isPaused = false;
        }

        private void Update()
        {
            if (agent != null && !agent.pathPending && !isPaused)
            {
                if (agent.remainingDistance <= agent.stoppingDistance)
                {
                    if (!agent.hasPath || agent.velocity.sqrMagnitude == 0f)
                    {
                        Destroy(this);
                    }
                }
            }
        }
    }
}
```

```json
{
    "name": "Run pathfinding behaviour",
    "description": "Runs a pathfinding behaviour to navigate to a position",
    "endCriteria": [...],
    "botAction": {
        "type": "Behaviour",
        "data": {
            "behaviourFullName": "RG.Example.NavigateToPosition",
            "maxRuntimeSeconds": 30.0
        }
    }
}
```

### MonkeyBot

The `MonkeyBot` action type runs a chaos monkey bot for a given amount of time. This can be useful for testing the robustness of your game.
This Monkey Bot will randomly select actions to perform from a list of possible actions, and will perform them at a given interval. These

> action are found via a code analysis of the game, which must be run before you use this segment type.

Reviewer 3: ```suggestion
actions are found via a code analysis of the game, which must be run before you use this segment type.
```
